Migrate blueprint feature flag to core feature system

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/55114

This PR migrates the Blueprint feature flag from WCA to the core feature system.

> * List Blueprint in Settings > Advanced > Features
> * Use the name "Blueprint (beta)"
> * Use ID/keys of "blueprint" to ensure we can remove the beta label in the future.
> * Use the description "Enable blueprint to import and export settings in bulk"
> * Enable the feature by default


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site with this branch
2. Go to `Settings > Advanced > Features`
3. See "Blueprint (beta)" in the list and it's disabled by default
4. Turn on the feature and go to `Settings > Advanced > Blueprint`
5. See the new Blueprint page
6. Try to export settings, and it should work
7. For devs, you should see blueprint cli commands available by running `wp wc blueprint`
8. Disable the feature, and blueprint page should not be accessible


<img width="1038" alt="Screenshot 2025-02-05 at 16 01 24" src="https://github.com/user-attachments/assets/3e014e12-2532-47ac-9d58-e6ec270404ce" />

```bash
> wp wc blueprint
usage: wp wc blueprint export <save-to> [--steps=<steps>] [--format=<format>]
   or: wp wc blueprint import <schema-path> [--show-messages=<show-messages>]

See 'wp help wc blueprint <command>' for more information on a specific command.
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Migrate blueprint feature flag to core feature system

- Replace direct feature config checks with `FeaturesUtil::feature_is_enabled()`
- Remove hardcoded blueprint feature flags from config files
- Update feature initialization in Features class
- Add blueprint feature definition in FeaturesController

chihsuan committed:
Add changelog

chihsuan committed:
Mark blueprint feature as legacy for compatibility handling

Add a comment explaining the rationale for marking the blueprint feature as a legacy feature, despite not being a true legacy feature. This change ensures consistent compatibility checking similar to other features like order attribution.

chihsuan commented on plugins/woocommerce/src/Admin/Features/Features.php:
```diff
@@ -146,6 +147,10 @@ public static function load_features() {
+
+		if ( FeaturesUtil::feature_is_enabled( 'blueprint' ) ) {
+			new \Automattic\WooCommerce\Admin\Features\Blueprint\Init();
+		}
```
WCA feature flag system automatically initializes the feature class when the feature is enabled but core system doesn't do this so we need to manually initialize the feature class. I think we should initialize the feature class in the same place as WCA does so it's easier to maintain. Open to suggestions.

adrianduffell replied:
Another option I see is to initialize it with the [DI container in the WooCommerce class](https://github.com/woocommerce/woocommerce/blob/ccc6eb86c76bafdb642f09c5f694e801dee9e414/plugins/woocommerce/includes/class-woocommerce.php#L314-L315). I'm good with what you have though.

chihsuan replied:
Thanks for the suggestion! 🙌   On second thought, I think it's better to keep it as is so we only initialize it in the admin context.

chihsuan committed:
Fix lint

chihsuan committed:
Fix tests

adrianduffell commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -312,6 +312,26 @@ private function get_feature_definitions() {
 					),
 				),
+				'blueprint'              => array(
+					'name'               => __( 'Blueprint (Beta)', 'woocommerce' ),
```
```suggestion
					'name'               => __( 'Blueprint (beta)', 'woocommerce' ),
```

Apologies, could we make this sentence case?

adrianduffell reviewed (approved):
Thanks @chihsuan, LGTM and tested well! I left a couple suggestions.

Pre-approving! 🚀

chihsuan committed:
Update plugins/woocommerce/src/Internal/Features/FeaturesController.php

Co-authored-by: Adrian Duffell <9312929+adrianduffell@users.noreply.github.com>

ilyasfoo commented on plugins/woocommerce/includes/class-wc-cli.php:
```diff
@@ -65,8 +66,7 @@ private function maybe_include_blueprint_cli() {
 
-		$features = wc_admin_get_feature_config();
-		if ( isset( $features['blueprint'] ) ) {
+		if ( FeaturesUtil::feature_is_enabled( 'blueprint' ) ) {
```
Previously this checks existence of the feature, but now it's checking if it's enabled. I think the change intended, but just making sure

ilyasfoo replied:
May not be related to this PR, but is it intended for the cli command to be available despite disabling the feature?

<img width="1051" alt="image" src="https://github.com/user-attachments/assets/56abd50c-81a5-403c-9202-9d160f6f9768" />

chihsuan replied:
> May not be related to this PR, but is it intended for the cli command to be available despite disabling the feature?

That's a good question! I didn't check this. I thought it should be disabled when feature flag is off. @moon0326 Could you confirm if this is the intended behavior?

adrianduffell replied:
Wonder if we should actually put the CLI command under a separate feature flag (non-UI) and release it later on? I'm thinking the command adds a bit too much for us to handle in MVP 🤔

adrianduffell replied:
>  I thought it should be disabled when feature flag is off

@chihsuan, yeah that's correct - it should be controlled by a feature flag

chihsuan replied:
Thanks, I've fixed conditional checks for Blueprint CLI in 67bf137e3e92ef63dcf2f9ef1d446a5caf9b01b0.

> Wonder if we should actually put the CLI command under a separate feature flag (non-UI) and release it later on?

IMO it's fine to release it together because it seems that CLI and Blueprint REST API share most of the code and CLI only provides two commands (import/export). cc @moon0326 since you have the best understanding of the code. 🙏 

I'll merge this PR first and we can do it in a follow-up PR if we want to change this.

ilyasfoo reviewed (approved):
Tests well! I have minor comments but pre-approving

ilyasfoo commented on plugins/woocommerce/client/admin/config/core.json:
```diff
@@ -38,7 +38,6 @@
 		"async-product-editor-category-field": false,
 		"launch-your-store": true,
 		"product-editor-template-system": false,
-		"blueprint": false,
```
@chihsuan  1 more thing to confirm, previously blueprint is disabled by default in production build, but now I assume it's not (essentially launching the feature in the next version)?

chihsuan replied:
Yes, I think this is what we want to do according to the issue description.  @adrianduffell To confirm, do we still want to turn it on by default? As you just mentioned we may push back the blueprints MVP date.

adrianduffell replied:
@chihsuan It's looking likely that we will push the release date back, I'll know later today. If needed, You could merge it with it disabled by default for now.

chihsuan replied:
Okay, I'll update the PR to disable it by default. Thanks for confirming!

chihsuan replied:
Updated in 645467c4fbb04f6371f94a21632152d4b01b2cab.

chihsuan committed:
Disable blueprint by default

chihsuan committed:
Update Blueprint CLI loading conditions

- Modify CLI initialization to check Blueprint feature flag before loading
- Remove unnecessary feature config inclusion
- Improve conditional checks for Blueprint CLI registration
